Port index_copy from TH to ATen

lezcano opened the pull request:
The design of the `TensorIterator` was similar to that in https://github.com/pytorch/pytorch/pull/50578

Resolves https://github.com/pytorch/pytorch/issues/24670
Resolves https://github.com/pytorch/pytorch/issues/24523

Timings:
<details>
<summary>Script</summary>

```python
from IPython import get_ipython
import torch

torch.manual_seed(13)
torch.set_num_threads(1)

ipython = get_ipython()

cpu = torch.device('cpu')
cuda = torch.device('cuda')


def run_test(ndims, size, index_len, device):
    print(f"ndims: {ndims}, tensor_size: {size}, index_len: {index_len}, device: {device}")

    x = torch.rand(*([size] * ndims), device=device)
    index = torch.randint(size, (index_len,), dtype=torch.long, device=device)
    for d in range(ndims):
        shape_t = [size] * d + [index_len] + [size] * (ndims - d - 1)
        t = torch.rand(*shape_t, device=device)
        command = "x.index_copy(d, index, t)"
        if device == cuda:
            command = command + "; torch.cuda.synchronize()"
        ipython.magic(f"timeit {command}")
    print()


run_test(3, 700, 10, cpu)
run_test(3, 700, 100, cpu)
run_test(3, 700, 700, cpu)
run_test(2, 10000, 10000, cpu)

run_test(3, 700, 10, cuda)
run_test(3, 700, 100, cuda)
run_test(3, 700, 700, cuda)
run_test(2, 10000, 10000, cuda)
```

</details>

<details>
<summary>CPU ATen</summary>

```
ndims: 3, tensor_size: 700, index_len: 10, device: cpu                                                                                                                                                          
327 ms ± 309 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)                                                                                                                                              
329 ms ± 456 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)                                                                                                                                              
378 ms ± 1.44 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)                                                                                                                                             
                                                                                                                                                                                                                
ndims: 3, tensor_size: 700, index_len: 100, device: cpu                                                                                                                                                         
348 ms ± 1.52 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)                                                                                                                                             
359 ms ± 330 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)                                                                                                                                              
526 ms ± 686 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)                                                                                                                                              
                                                                                                                                                                                                                
ndims: 3, tensor_size: 700, index_len: 700, device: cpu                                                                                                                                                         
560 ms ± 19 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)                                                                                                                                               
552 ms ± 2.61 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)                                                                                                                                             
932 ms ± 2.52 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)                                                                                                                                             
                                                                                                                                                                                                                
ndims: 2, tensor_size: 10000, index_len: 10000, device: cpu                                                                                                                                                     
163 ms ± 5.05 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)                                                                                                                                           
302 ms ± 5.75 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```                                                                                                                                           
</details>

<details>
<summary>CUDA ATen</summary>

```
ndims: 3, tensor_size: 700, index_len: 10, device: cuda
9.63 ms ± 441 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
9.65 ms ± 230 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
12.4 ms ± 881 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)

ndims: 3, tensor_size: 700, index_len: 100, device: cuda
10.8 ms ± 1.51 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
11 ms ± 417 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
21.2 ms ± 18.2 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

ndims: 3, tensor_size: 700, index_len: 700, device: cuda
19 ms ± 4.42 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
17.8 ms ± 493 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
25.8 ms ± 1.22 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

ndims: 2, tensor_size: 10000, index_len: 10000, device: cuda
5.59 ms ± 109 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
10 ms ± 25.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

</details>

<details>
<summary>CPU TH</summary>

```
ndims: 3, tensor_size: 700, index_len: 10, device: cpu
333 ms ± 2.42 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
327 ms ± 1.04 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
366 ms ± 753 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

ndims: 3, tensor_size: 700, index_len: 100, device: cpu
336 ms ± 1.24 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
345 ms ± 914 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
884 ms ± 4.32 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

ndims: 3, tensor_size: 700, index_len: 700, device: cpu
441 ms ± 3.58 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
514 ms ± 1.17 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
7.46 s ± 6.46 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

ndims: 2, tensor_size: 10000, index_len: 10000, device: cpu
141 ms ± 233 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
1.13 s ± 855 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

</details>


<details>
<summary>CUDA TH</summary>

```
ndims: 3, tensor_size: 700, index_len: 10, device: cuda
9.64 ms ± 390 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
9.68 ms ± 3.26 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
13.9 ms ± 928 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)

ndims: 3, tensor_size: 700, index_len: 100, device: cuda
11.6 ms ± 1.38 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
12.1 ms ± 3.72 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
30.3 ms ± 27.2 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

ndims: 3, tensor_size: 700, index_len: 700, device: cuda
27.2 ms ± 19.8 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
30.6 ms ± 43.6 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
146 ms ± 204 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

ndims: 2, tensor_size: 10000, index_len: 10000, device: cuda
6.5 ms ± 3.99 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
64.7 ms ± 55.5 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

</details>

According to these we see a slight performance improvement across both CPU and GPU.


cc: @nikitaved

lezcano committed:
Initial implementation of index_copy_. Tests pass

lezcano committed:
Implemented index stride optimisation

Implemented CUDA backend. Fixed the signature of the cpu version

Delete old TH files

Remove two last indexCopy appearances

Fix a side-effect that made some tests fail.
Add OpInfo tests.
Add index_copy to the lists of functions whose derivatives work with complex numbers

Simpler test. Comment out negative indices

pep8

More thorough testing
index_copy does not support negative indices now
Tested same device and dtype explicitly now

Simpler test

More tests. Mention function name in errors.

Handles the empty types correctly
Tests all the possible empty types

Lint tests

corrected size

typing

Errata test

Delete legacy test
Separate problematic XLA test and document it

Added issues

Delete test
Update comment on the derivative

Correct keyword in test

Trimmed test

nikitaved commented on aten/src/ATen/native/cuda/IndexKernel.cu:
```diff
@@ -192,6 +192,45 @@ void index_fill_kernel_impl(
+  auto loop = [=]C10_DEVICE(int i) {
+    auto offsets = offset_calc.get(i);
+
+    auto* __restrict__ self_data = reinterpret_cast<scalar_t*>(self_ptr + offsets[0]);
+    auto idx = *reinterpret_cast<int64_t*>(idx_ptr + offsets[1]);
+    auto* __restrict__ source_data = reinterpret_cast<scalar_t*>(source_ptr + offsets[2]);
```
What if `self` and `source` overlap? Should we actually allow memory overlap, or is it better to forbid that?
Probably makes sense to have two kernels: with and without memory overlap to enable that restrict modifier.

Lezcano replied:
At the moment this is not supported. See the test in `test_torch.py`
```python
    @onlyOnCPUAndCUDA
    def test_index_copy_mem_overlap(self, device):
        x = torch.rand((1,), device=device).expand((6,))
        y = torch.rand((6,), device=device)
        ind = torch.tensor([2, 1, 0], device=device)
        value = torch.rand((3,), device=device)
        with self.assertRaisesRegex(RuntimeError, 'unsupported operation'):
            x.index_copy_(0, ind, value)
        with self.assertRaisesRegex(RuntimeError, 'unsupported operation'):
            y.index_copy_(0, ind, y[:3])
        with self.assertRaisesRegex(RuntimeError, 'unsupported operation'):
            ind.index_copy_(0, ind, ind.clone())
        with self.assertRaisesRegex(RuntimeError, 'unsupported operation'):
            ind.index_copy_(0, ind.clone(), ind)
```

nikitaved commented on aten/src/ATen/LegacyTHFunctionsCPU.cpp:
```diff
@@ -165,72 +165,6 @@ Tensor _th_nonzero(const Tensor & self) {
     }
     return result;
 }
-Tensor & _th_index_copy_(Tensor & self, int64_t dim, const Tensor & index, const Tensor & source) {
-    // DeviceGuard omitted
-    auto dispatch_scalar_type = infer_scalar_type(self);
```
check out `git grep indexCopy` and remove these entries as well.

nikitaved replied:
Ooops, sorry, did not update the page to see the rest having been already removed.

nikitaved commented:
@Lezcano , could you please remove legacy tests and implement them with OpInfo? You could have a look at how `index_fill` is doing that.

Lezcano commented:
Just implemented the `OpInfo` tests. This PR should be ready to be reviewed.

nikitaved commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -704,6 +704,30 @@ def sample_inputs_index_fill(op_info, device, dtype, requires_grad):
+    # non-contiguous input
+    t01 = t.transpose(0, 1)
+    t02 = t.transpose(0, 2)
+    t12 = t.transpose(1, 2)
+    idx = index_variable(M, S, device=device)
+    # non-contiguous index
```
Hmm, shouldn't index be 1d?

nikitaved commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -411,11 +418,65 @@ Tensor & index_copy_(Tensor & self, int64_t dim, const Tensor & index, const Ten
+  // Handle the case when `self` is 0-dim
+  Tensor self_nonzero = self.dim() == 0 ? self.unsqueeze(-1) : self;
+
+  // The only different between the following  tensor iterator and that of index_fill_ is that
+  // this one has also source as an input. Maybe refactor?
```
That is actually a good question. The issue is that it turns out `TensorIteratorConfig` is not copyable nor copy-assignable. So, it looks like only the restriding part could be shared. It is possible to write the same kernel, however, with the same kernel callback that returns a pointer to the output, source and index.

Lezcano commented:
I put the [WIP] check, as the backward `OpInfo` tests are failing. In particular, one of the missing parts is for `index_select` to support negative indices, but there are a couple more things that fail. I will look into them now.

Lezcano commented:
So, I am getting three kinds of errors:

The first one is coming from the fact that the backwards uses `index_select`, and `index_select` does not accept negative indices. I do not know whether there is a reason for this behaviour. The good solution for this would be, of course, to add this support for `index_select` (and all the `index_` operations really), as some of them support it (see `index_fill`). This would require some amount of work.

The second kind of errors that I get are of the following form:
```python
grad = None                                          
    def retain_grad_hook(grad):
        var = weak_self()
        if var is None:                              
            return                         
        if var._grad is None: 
            if grad.is_sparse:          
                var._grad = grad.clone()             
            else:                                                                      
                var._grad = grad.clone(memory_format=torch.contiguous_format)                             
        else:                              
>           var._grad = var._grad + grad                                               
E           TypeError: unsupported operand type(s) for +: 'Tensor' and 'NoneType'                         
torch/tensor.py:363: TypeError                                                         
________________ TestGradientsCPU.test_fn_grad_index_copy_cpu_float64 _________________ 
```
in the tests
```
FAILED test/test_ops.py::TestGradientsCPU::test_fn_grad_index_copy_cpu_complex128
FAILED test/test_ops.py::TestGradientsCPU::test_fn_grad_index_copy_cpu_float64                                                                                                                    
FAILED test/test_ops.py::TestGradientsCUDA::test_fn_grad_index_copy_cuda_complex128                       
FAILED test/test_ops.py::TestGradientsCUDA::test_fn_grad_index_copy_cuda_float64 
```
The implementation in `master` throws exactly the same errors in these tests. Should I care about these?

The third kind of errors are of the form:
```python
method_name = 'index_copy'                                                                                                                                                                                          
args = (tensor([[[-4,  8, -3, -9,  7],                                                                                                                                                                              
         [ 1,  5,  9, -6, -3],                                                                                                                                                                                      
         [-6, -3, -8, -8, -6],                                                                                                                                                                                      
         [-1, -3,  0,  6..., -9, -3, -8,  4],                                                                                                                                                                       
         [-2,  9,  1,  1,  4],                                                                                                                                                                                      
         [-9, -9, -6,  8,  4]]], device='cuda:0', dtype=torch.int32))                                                                                                                                               
kwargs = {}                                                                                                                                                                                                         
                                                                                                                                                                                                                    
    def check_alias_annotation(method_name, args, kwargs, *, aten_name, func_type='method'):                                                                                                                        
        formals, tensors, actuals = get_script_args(args)                                                                                                                                                           
        call = get_call(method_name, func_type, actuals, kwargs)                                                                                                                                                    
        script = script_template.format(', '.join(formals), call)                                                                                                                                                   
        CU = torch.jit.CompilationUnit(script)                                                                                                                                                                      
>       torch._C._jit_check_alias_annotation(CU.the_method.graph, tuple(tensors), aten_name)                                                                                                                        
E       RuntimeError: !lhs.isAliasOf(rhs) INTERNAL ASSERT FAILED at "../torch/csrc/jit/passes/utils/check_alias_annotation.cpp":92, please report a bug to PyTorch.                                                 
                                                                                                                                                                                                                    
torch/testing/_internal/jit_metaprogramming_utils.py:478: RuntimeError
```
These errors are thrown in:
```
FAILED test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_index_copy_cpu_float32
FAILED test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_index_copy_cpu_int32
FAILED test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_index_copy_cuda_float32
FAILED test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_index_copy_cuda_int32
```
These errors are also thrown in the implementation in `master`.

How should I go about each of these errors?
CC: @nikitaved @ngimel

mruberry commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -385,8 +386,14 @@ Tensor & index_copy_(Tensor & self, int64_t dim, const Tensor & index, const Ten
   }
 
+  // Should this be TORCH_CHECK?
   TORCH_CHECK_INDEX(index.scalar_type() == ScalarType::Long, "index_copy_(): Expected LongTensor for index");
```
Yep, this should be TORCH_CHECK

Also "LongTensor" is antiquated. Expected a "long tensor for index, but got [dtype]" is more modern.

mruberry commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -385,8 +386,14 @@ Tensor & index_copy_(Tensor & self, int64_t dim, const Tensor & index, const Ten
 
+  // Added in accordance with index_fill_
+  if (!self.is_complex() && source.is_complex()) {
+    TORCH_CHECK(false, "index_copy_(): Converting complex source Tensor to non-complex type is forbidden");
```
"Tensor" -> "tensor"

"is forbidden" -> "is not supported"

mruberry replied:
This doesn't seem sufficient. Shouldn't the src and dst tensors be checked for having the SAME dtype?

Another question, is it ever validated that all 3 tensors are on the same device?

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -704,6 +704,37 @@ def sample_inputs_index_fill(op_info, device, dtype, requires_grad):
+    t02 = t.transpose(0, 2)
+    t12 = t.transpose(1, 2)
+    # non-contiguous input
+    s01 = t.transpose(0, 1)
```
This is probably the source of the gradcheck failures. It looks like you mean to use `s.transpose` here? Gradcheck has an oddity where 

```
gradcheck(torch.index_copy, (a, 0, torch.tensor((0, 1)), a.T))
```

will fail, even though computing the gradient in eager mode will succeed. cc @albanD

mruberry replied:
This MAY be the source of the jit failure, too, although the operator is jittable in this case on Colab:

```
def foo(t):
  return torch.index_copy(t, 0, torch.tensor((1, 0)), t.T)
scripted_foo = torch.jit.script(foo)
out = scripted_foo(a)
```

Lezcano replied:
You are completely right! Sorry for this, this was my first PR on the C++ side, and after looking up the JIT errors, it seemed like it could have been caused by some internals.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -704,6 +704,37 @@ def sample_inputs_index_fill(op_info, device, dtype, requires_grad):
+            for source in [s, s01, s02, s12]:
+                samples.append(SampleInput((tensor, d, idx, source)))
+                # Backwards is implemented with `index_select`, which does not support negative indices at the moment
+                # samples.append(SampleInput((tensor, d, -idx - 1, source)))
```
The forward of this function also did not used to support negative indices, right? We probably don't want to change that behavior (and should add a test for it). For example, 

```
torch.index_copy(a, 0, torch.tensor((0, -1)), b)
```

used to throw 

```
RuntimeError: invalid argument 3: out of range at /pytorch/aten/src/TH/generic/THTensor.cpp:245
```

which is an odd error message that should be improved, but an error for negative indices.

mruberry replied:
@ngimel Notes that a follow-up PR might enable negative indices, but the issue with index_select needs to be dealt with.

Lezcano replied:
I removed now the support for negative indices following your advice

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -704,6 +704,37 @@ def sample_inputs_index_fill(op_info, device, dtype, requires_grad):
+    # non-contiguous index
+    idx_nonctg = torch.empty_strided((S,), (2,), device=device, dtype=torch.int64)
+    idx_nonctg.copy_(idx)
+    for d in range(t.dim()):
```
Style nit: one for loop over the product of these iterables

mruberry reviewed (commented):
Hey @Lezcano,

Overall looks good. @ngimel and I took some time to analyze your errors and made a few suggestions. Looking forward to reviewing your analysis!

Lezcano commented:
Hi @mruberry
I have corrected all the things that you mentioned. This PR should be ready :)

mruberry commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -411,11 +416,65 @@ Tensor & index_copy_(Tensor & self, int64_t dim, const Tensor & index, const Ten
 
+Tensor & _index_copy_impl_(Tensor & self, int64_t dim, const Tensor & index, const Tensor & source) {
+  // Handle the case when `self` is 0-dim
+  Tensor self_nonzero = self.dim() == 0 ? self.unsqueeze(-1) : self;
```
@ngimel: this seems odd. What's going on here? Won't this result in an out-of-place operation since self_nonzero will be a different tensor? More precisely, while the data might be manipulated in place, self_nonzero would be a distinct tensor object from self.

ngimel replied:
nevermind, `self` is ultimately returned, so all's good.

mruberry commented on aten/src/ATen/native/cpu/IndexKernel.cpp:
```diff
@@ -200,6 +200,69 @@ void index_fill_kernel(
+      auto* index_data_bytes = data[1];
+      auto* source_data_bytes = data[2];
+      auto idx = *reinterpret_cast<int64_t*>(index_data_bytes);
+      if (idx < -self_dim_size || idx >= self_dim_size) {
```
Doesn't this index still need to be fixed? 

Also, these error messages should start with the name of the operation (index_copy_, like in the above error messages).

Lezcano replied:
Absolutely. I have fixed these two things now.

mruberry commented on aten/src/ATen/native/cuda/IndexKernel.cu:
```diff
@@ -192,6 +192,42 @@ void index_fill_kernel_impl(
+    auto* __restrict__ self_data = reinterpret_cast<scalar_t*>(self_ptr + offsets[0]);
+    auto idx = *reinterpret_cast<int64_t*>(idx_ptr + offsets[1]);
+    auto* __restrict__ source_data = reinterpret_cast<scalar_t*>(source_ptr + offsets[2]);
+    CUDA_KERNEL_ASSERT(idx >= 0 && idx < self_dim_size && "index out of bounds");
```
Also add the op name here, like in the above error messages

mruberry commented on test/test_torch.py:
```diff
@@ -4385,37 +4385,40 @@ def test_narrow_empty(self, device):
+
+        def make_arg(batch_sizes, n, dim, contig):
+            size_arg = batch_sizes[:dim] + (n,) + batch_sizes[dim:]
+            arg = make_tensor(size_arg, device, dtype, low=None, high=None)
```
Use the "discontiguous" kwarg in make_tensor:

https://github.com/pytorch/pytorch/blob/1d6bd157902d4b1347a5d03122d02b407658e263/torch/testing/_internal/common_utils.py#L1531

instead of make_non_contiguous()

mruberry commented on test/test_torch.py:
```diff
@@ -4385,37 +4385,40 @@ def test_narrow_empty(self, device):
+                        dest2[idx_dest] = src[idx_src]
+                    self.assertEqual(dest, dest2)
+
+    def test_errors_index_copy(self, device):
```
The update to this test is good but it should validate the 3 cases in the kernels (2 on CPU, 1 on CUDA) that will throw index out of bound errors, too

mruberry commented on test/test_torch.py:
```diff
@@ -4385,37 +4385,40 @@ def test_narrow_empty(self, device):
+                    src = make_arg(other_sizes, num_copy, dim, dest_contig)
+                    idx = torch.randperm(num_dest, dtype=torch.int64, device=device).narrow(0, 0, num_copy)
+                    if not index_contig:
+                        idx = torch.testing.make_non_contiguous(idx)
```
Instead of calling this function, just make this tensor noncontiguous by copying these two lines from make_tensor:

```
https://github.com/pytorch/pytorch/blob/1d6bd157902d4b1347a5d03122d02b407658e263/torch/testing/_internal/common_utils.py#L1581
```

mruberry reviewed (commented):
Changes look good. Just a few more tweaks!

mruberry commented on test/test_torch.py:
```diff
@@ -4385,37 +4385,40 @@ def test_narrow_empty(self, device):
-        self.assertEqual(dest, torch.tensor([[True, True], [True, True]], device=device))
-
+    @dtypes(*torch.testing.get_all_dtypes())
+    def test_index_copy(self, device, dtype):
```
Is there a test case for correct 0-dim handling?

Lezcano replied:
These two test those cases:

https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_methods_invocations.py#L3328

mruberry replied:
Please add those test cases to the sample inputs and remove the method_tests() entries.

They'll still need an explicit test here, though, because sample inputs only validate that autograd and the jit work correctly. They don't check that the function computes the correct value in those cases.

Lezcano replied:
Now it tests all the empty cases correctly. I deleted these two legacy tests.
You'll be pleased to know (or not really) that this PR is also affected by the weird issue #49707

ngimel replied:
@Lezcano that issue has an easy fix https://github.com/pytorch/pytorch/issues/49707#issuecomment-788140578, but I don't know if it's better to make it in a separate PR or here.

Lezcano replied:
I would be happy to add the fix to this PR. What do you think @mruberry ?
Other possibility would be to land this PR and then have another PR with the fix proposed by @ngimel + uncomment the relevan tests for both `index_copy` and `index_select`.

ngimel replied:
Let's have a fix in a follow-up PR then. Thanks @Lezcano !

Lezcano commented:
I have corrected all the points that you brought up. The only thing that might be left for you to check is the tests that test the zero dimensional case. Do you want me to add those to the `OpInfo`  tests and remove the legacy tests?

Lezcano commented:
@mruberry I have added the tests for the scalar cases both in the forward in `test_torch.py` and the backward in the `OpInfo` as we discussed. All the combinations were not properly tested nor working before, now they all pass.

mruberry commented:
XLA failure looks real, other failures appear to be in base. I wonder if the XLA issue is a bug in XLA?

@ngimel and I will try and take another look soon, @Lezcano.

anjali411 commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -385,7 +386,11 @@ Tensor & index_copy_(Tensor & self, int64_t dim, const Tensor & index, const Ten
 
-  TORCH_CHECK_INDEX(index.scalar_type() == ScalarType::Long, "index_copy_(): Expected LongTensor for index");
+  TORCH_CHECK(index.scalar_type() == ScalarType::Long, "index_copy_(): Expected a long tensor for index, but got ", index.scalar_type())
+  TORCH_CHECK(self.scalar_type() == source.scalar_type(), "index_copy_(): self and source expected to have the same dtype, but got (self) ", self.scalar_type(), " and (source) ", source.scalar_type());
```
I wonder if we should relax this condition to allow copying ints to floating point tensors and floats to complex tensors etc. 
```
>>> t = torch.tensor([[1, 2, 3], [4, 5, 6], [7, 8, 9]], dtype=torch.int)
>>> index = torch.tensor([0, 4, 2])
>>> x.index_copy_(0, index, t)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Expected object of scalar type Float but got scalar type Int for argument #4 'source' in call to _th_index_copy_
```
I propose relaxing the existing condition, because we do allow similar behavior in `torch.Tensor.copy_()`:
```
>>> x=torch.zeros(2,3, dtype=torch.int)
>>> y=torch.randn(2,3)
>>> y.copy_(x)
tensor([[0., 0., 0.],
        [0., 0., 0.]])
```
cc. @ngimel

Lezcano replied:
I think this makes sense. Relaxing this to `at::can_cast(source, self)` would also make the preconditions on these operations consistent with the `out=` rules. At the moment, the `out=` version of `torch.take` has this `self.scalar_type() == source.scalar_type()` kind of check, which is more restrictive than the `out=` operations in `torch.linalg`.

The only problem is that, for this to be consistent, it would have to be implemented across all the methods for advanced indexing, which will take some time.

mruberry replied:
Maybe a good issue to follow-up with, @anjali411? I'd prefer this PR not add or change functionality since it's principally about porting the operation (although fixing bugs would be fine).

anjali411 replied:
@mruberry yeah that makes sense! @Lezcano could you file an issue for that?

Lezcano replied:
Done in https://github.com/pytorch/pytorch/issues/53257

mruberry commented on test/test_torch.py:
```diff
@@ -4385,37 +4385,70 @@ def test_narrow_empty(self, device):
+                    make_tensor(size_s, dtype=dtype, device=device, low=None, high=None))
+                   for size_t, size_i, size_s in product([(), (1,)], repeat=3))
+        for target, idx, source in scalars:
+            target.index_copy_(0, idx, source)
```
Looks like this part of the test is causing PyTorch/XLA trouble.

Would you take this segment of the test and put it into its own test, then mark that test as @onlyOnCPUAndCUDA. Separately, would you file an issue with the PyTorch/XLA team (https://github.com/pytorch/xla) about this failure, and put the link to that issue in the comment above the skip?

Lezcano replied:
Issue filed: https://github.com/pytorch/pytorch/issues/53256

mruberry commented on tools/autograd/derivatives.yaml:
```diff
@@ -562,7 +562,7 @@
 - name: index_copy_(Tensor(a!) self, int dim, Tensor index, Tensor source) -> Tensor(a!)
   self: grad.clone().index_fill_(dim, index, 0)
-  source: grad.index_select(dim, index).expand_as(source)
+  source: "source.dim() > 0 ? grad.index_select(dim, index).expand_as(source) : grad.index_select(dim, index.squeeze(0))"
```
Please add a comment above this gradient explaining this source formula.

@ngimel asks, "Is the proper fix for this updating index_select()? If so, let's file an issue for that."

Lezcano replied:
So, what's happening here is that, for a 0-dimensional source, `expand_as(source)` fails.  I believe that it should not fail if the input on which it is being applied is of shape `()` or `(1,)`. This might be an edge-case for  `expand_as` that might not be tested, and it might not be implemented correctly. Should I file an issue about this, or is this an expected behaviour of `expand_as`?

mruberry replied:
Thank you for this explanation. An issue would be a great, and a comment linking the issue here would be awesome.

Lezcano replied:
Filed here: https://github.com/pytorch/pytorch/issues/53351

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -3349,8 +3391,6 @@ def method_tests():
         ('index_copy', (S, S), (0, index_perm_variable(2, S), (2, S)), 'dim', (), [0]),
```
Can these other index_copy entries in method_tests() be removed to? Do new sample inputs need to be created for them?

Lezcano replied:
The first one I have removed. The second one is not so clear how to implement it with  `OpInfo`, as it has an `expectedAlertNondeterministic('index_copy')`. How can that be translated to an `OpInfo` test?

mruberry reviewed (commented):
Hey @Lezcano! This looks very good and is very close. Just a few final issues for you review. Most notably:

- we suspect the PyTorch/XLA failure you're hitting is a failure in PyTorch/XLA, so we've suggested skipping that part of the test on PyTorch/XLA and notifying them of the issue
- we want to kill the remaining method_tests() entries for the function

After that we should be good to go!

Lezcano commented:
@mruberry 
I have made the changes that you mentioned. There are two details remaining though:
- The workaround for the derivative due to the problematic empty case for `resize_as`
- The legacy test containing `expectedAlertNondeterministic('index_copy')`

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -3338,11 +3380,8 @@ def method_tests():
         ('index_add', (S, S), (0, index_variable(2, S), (2, S)), 'alert_nondeterministic', (), [0],
             [expectedAlertNondeterministic('index_add_cuda_', 'cuda')]),
-        ('index_copy', (S, S), (0, index_perm_variable(2, S), (2, S)), 'dim', (), [0]),
         ('index_copy', (S, S), (0, index_perm_variable(2, S), (2, S)), 'dim_alert_nondeterministic', (), [0],
```
I think it's OK to add this case without the expectedAlertNondeterministic as long as we file an issue that we want to create an OpInfo-based mechanism for testing nondeterminism and link that issue in a comment above the index copy sample inputs function.

mruberry commented:
> @mruberry
> I have made the changes that you mentioned. There are two details remaining though:
> 
> * The workaround for the derivative due to the problematic empty case for `resize_as`
> * The legacy test containing `expectedAlertNondeterministic('index_copy')`

We made inline comments for both these cases. We think filing an issue and linking that issue in a comment would be a great solution.

Lezcano commented:
@mruberry I just filed the issues and linked them appropriately in the code.

Lezcano commented:
I just deleted the test with the alert non deterministic behaviour (I forgot to do so in the last commit) and updated the comment on the derivative of the function as per the offline discussion with Mike.

This is ready for a (hopefully last) review

cc @mruberry @ngimel

mattip commented:
There is a merge conflict in this PR

Lezcano commented:
Thanks for pointing it out. It should hopefully be solved now. 
(ping @mruberry )

mruberry commented:
ASAN failure looks unrelated, although this does need to resolve a merge conflict. @ngimel and I will take a final look soon, @Lezcano.

Lezcano commented:
Merge conflict solved. It's difficult to keep up with the rhythm at which ports of backwards to complex are being merged :)

mruberry commented:
> Merge conflict solved. It's difficult to keep up with the rhythm at which ports of backwards to complex are being merged :)

You're telling me!

mruberry reviewed (approved):
Nice job, @Lezcano! @ngimel and I both approve!

mruberry commented:
Unfortunately looks like this picked up several merge conflicts preventing its merging. Would you please rebase this, @Lezcano?

Lezcano commented:
I just did.  @mruberry

mruberry commented:
Unfortunately more merge conflicts, @Lezcano, and I don't have write perms to resolve them, sorry :(

Would you fix them and ping me?

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -3126,6 +3159,16 @@ def _make_tensor_helper(shape, low=None, high=None):
            sample_inputs_func=sample_inputs_index_fill),
+    OpInfo('index_copy',
+           dtypes=all_types_and_complex_and(torch.bool, torch.float16, torch.bfloat16),
+           test_inplace_grad=False,
```
supports_inplace_grad=False

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -3126,6 +3159,16 @@ def _make_tensor_helper(shape, low=None, high=None):
+           skips=(
+               # https://github.com/pytorch/pytorch/issues/49707
+               # Fails on cpu throwing: RuntimeError: "index_select" not implemented for 'Half' / 'BFloat16'
+               SkipInfo('TestCommon', 'test_variant_consistency_jit', dtypes=[torch.float16, torch.bfloat16]),
```
This skip can be removed but a new skip is likely needed for test_supported_backward

Lezcano replied:
In fact, https://github.com/pytorch/pytorch/pull/53898 corrected this problem, so now we should be able to get rid of this skip altogether.

lezcano committed:
Update tests

lezcano committed:
lint
